src/lib/environment.ts

```diff
@@ -48,40 +49,75 @@ export class Environment<StateT, ActionT extends Action> {
     return selector(this.store.getState(), ...args);
   };
 
-  public run<T, P extends any[]>(f: (...params: P) => T, ...params: P): T;
-  public run<T, P extends any[]>(effect: Effect<StateT, ActionT, P, T>, ...params: P): T;
-  public run<T, P extends any[]>(funcOrEffect: Effect<StateT, ActionT, P, T> | ((...params: P) => T), ...params: P): T {
+  /**
+   * Calls `f` with the supplied params
+   *
+   * In case the return type is not correctly inferred, wrap the function in an BoundEffect
+   *
+   * @param f
+   * @param params
+   */
+  public call = <T, P extends any[]>(f: (...params: P) => T, ...params: P): T => {
     if (this.cancellationToken && this.cancellationToken.canceled) {
       throw new SagaCancelledError(`Saga has been cancelled`);
     }
 
-    if (typeof funcOrEffect === 'function') {
-      return funcOrEffect(...params);
-    } else {
-      return funcOrEffect.func(this, ...params);
-    }
-
-    // return f(...params);
-  }
+    return f(...params);
+  };
 
   public take = async <Payload>(actionCreator: ActionCreator<Payload>): Promise<Payload> => {
     return (await this.waitForMessage(actionCreator)).payload;
   };
 
-  // public spawn = <P extends any[], T>(f: (env: Environment<StateT, ActionT>, ...args: P) => T, ...args: P): T => {
-  //   return f(this, ...args);
-  // };
+  /**
+   * Runs the given saga as an attached child.
+   * Cancelling the parent will also cancel the child at the next opportunity.
+   *
+   * @param effectOrEffectCreator
+   * @param params
+   */
+  public run<T, P extends any[]>(effect: BoundEffect<Environment<StateT, ActionT>, P, T>): T;
+  public run<T, P extends any[]>(effect: SagaFunc<StateT, ActionT, P, T>, ...params: P): T;
+  public run<P extends any[], T>(
+    effectOrEffectCreator: BoundEffect<Environment<StateT, ActionT>, P, T> | SagaFunc<StateT, ActionT, P, T>,
+    ...args: P
+  ): T {
+    // TODO: Add overload for (env: Env, ...) => T
```

Comment:
Missing check for cancellation

Reply:
🙏🏻. added.

I think for all the saga behavior (cancellation, forking) the tests are still severely lacking. Seems to work in practice for our limited use case, but we definitely need test cases for the edge cases.